processing/src/main/java/org/apache/druid/query/scan/ScanResultValueFramesIterable.java

```diff
@@ -0,0 +1,319 @@
+     */
+    final Function<RowSignature, Function<?, Object[]>> resultFormatMapper;
+
+    /**
+     * Accumulating the closers for all the resources created so far
+     */
+    final Closer closer = Closer.create();
+
+    /**
+     * Iterator from the scan result value's sequence, so that we can fetch the individual values. Closer registers the
+     * iterator so that we can clean up any resources held by the sequence's iterator, and prevent leaking
+     */
+    final ScanResultValueIterator resultSequenceIterator;
+
+    /**
+     * Either null, or points to the current non-empty cursor (and the cursor would point to the current row)
+     */
+    Cursor currentCursor = null;
+
+    /**
+     * Row signature of the current row
+     */
+    RowSignature currentRowSignature = null;
+
+
+    public ScanResultValueFramesIterator(
+        Sequence<ScanResultValue> resultSequence,
+        MemoryAllocatorFactory memoryAllocatorFactory,
+        boolean useNestedForUnknownTypes,
+        RowSignature defaultRowSignature,
+        Function<RowSignature, Function<?, Object[]>> resultFormatMapper
+    )
+    {
+      this.memoryAllocatorFactory = memoryAllocatorFactory;
+      this.useNestedForUnknownTypes = useNestedForUnknownTypes;
+      this.defaultRowSignature = defaultRowSignature;
+      this.resultFormatMapper = resultFormatMapper;
+      this.resultSequenceIterator = new ScanResultValueIterator(resultSequence);
+
+      closer.register(resultSequenceIterator);
+
+      // Makes sure that we run through all the empty scan result values at the beginning and are pointing to a valid
+      // row
+      populateCursor();
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+      return !done();
+    }
+
+    @Override
+    public FrameSignaturePair next()
+    {
+      if (!hasNext()) {
+        throw new NoSuchElementException("No more frames to produce. Call `hasNext()` before calling `next()`");
+      }
+
+      // It would ensure that the cursor and the currentRowSignature is populated properly before we
+      // start all the processing
+      populateCursor();
+      boolean firstRowWritten = false;
+      // While calling populateCursor() repeatedly, currentRowSignature might change. Therefore we store the signature
+      // with which we have written the frames
+      final RowSignature writtenSignature = currentRowSignature;
+      FrameWriterFactory frameWriterFactory = FrameWriters.makeFrameWriterFactory(
+          FrameType.COLUMNAR,
+          memoryAllocatorFactory,
+          currentRowSignature,
+          Collections.emptyList()
+      );
+      Frame frame;
+      try (final FrameWriter frameWriter = frameWriterFactory.newFrameWriter(new SettableCursorColumnSelectorFactory(
+          () -> currentCursor,
+          currentRowSignature
+      ))) {
+        while (populateCursor()) { // Do till we don't have any more rows, or the next row isn't compatible with the current row
```

Comment:
This should also mention that we need to create a new frame if the current frame is out of space no ?

Reply:
That is done by the `frameWriter.addSelection()`. I'll add that comment in the appropriate line.